packages/library-sync/src/commonMain/kotlin/io/realm/kotlin/mongodb/internal/MongoCollectionImpl.kt

```diff
@@ -0,0 +1,267 @@
+            projection?.let { put("project", it) }
+            sort?.let { put("sort", it) }
+        }
+        return call
+    }
+
+    @PublishedApi
+    internal suspend fun find(filter: BsonDocument? = null, projection: BsonDocument? = null, sort: BsonDocument? = null, limit: Long? = null): BsonValue =
+        call("find") {
+            filter?.let { put("query", it) }
+            projection?.let { put("project", it) }
+            sort?.let { put("sort", it) }
+            limit?.let { put("limit", BsonInt64(it)) }
+        }
+
+    @PublishedApi
+    internal suspend fun aggregate(pipeline: List<BsonDocument>): List<BsonValue> =
+        call("aggregate") { put("pipeline", BsonArray(pipeline)) }.asArray().toList()
+
+    @PublishedApi
+    internal suspend fun insertOne(document: BsonDocument): BsonValue =
+        call("insertOne") { put("document", document) }.asDocument()["insertedId"]!!
+
+    @PublishedApi
+    internal suspend fun insertMany(documents: List<BsonDocument>): List<BsonValue> =
+        call("insertMany") {
+            put("documents", BsonArray(documents))
+        }.asDocument()["insertedIds"]!!.asArray().toList()
+
+    @PublishedApi
+    internal suspend fun deleteOne(filter: BsonDocument): Boolean {
+        val deletedCountBson = call("deleteOne") {
+            put("query", filter)
+        }.asDocument()["deletedCount"]!!
+        val deletedCount = decodeFromBsonValue<Long>(deletedCountBson)
+        return when (deletedCount) {
+            0L -> false
+            1L -> true
+            else -> throw ServiceException("Unexpected response from deleteOne: deletedCount=$deletedCount")
+        }
+    }
+
+    @PublishedApi
+    internal suspend fun deleteMany(filter: BsonDocument): Long {
+        val deletedCountBson = call("deleteMany") {
+            put("query", filter)
+        }.asDocument()["deletedCount"]!!
+        return decodeFromBsonValue(deletedCountBson)
+    }
+
+    @PublishedApi
+    internal suspend fun updateOne(filter: BsonDocument, update: BsonDocument, upsert: Boolean = false): Pair<Boolean, BsonValue?> {
+        val response: BsonValue = call("updateOne") {
+            put("query", filter)
+            put("update", update)
+            put("upsert", BsonBoolean(upsert))
+        }
+        return response.asDocument().run {
+            val modifiedCount: Long? = get("modifiedCount")?.let { decodeFromBsonValue<Long>(it) }
+            val modified = when (modifiedCount) {
+                0L -> false
+                1L -> true
+                else -> throw ServiceException("Unexpected response from updateOne: modifiedCount=$modifiedCount")
+            }
+            modified to (get("upsertedId"))
+        }
+    }
+
+    @PublishedApi
+    internal suspend fun updateMany(filter: BsonDocument, update: BsonDocument, upsert: Boolean = false): Pair<Long, BsonValue?> {
+        val response = call("updateMany") {
+            put("query", filter)
```

Comment:
We could use constants instead of string values.

Reply:
Hmm. There is no magic interconnection between them from one method to another, so it is not necessarily something that needs to be commonized. I tried it out and it feels quite annoying and verbose, but I can do it, if you feel for it.